spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarPropertiesMapper.java

```diff
@@ -157,6 +162,14 @@ <T> void customizeProducerBuilder(ProducerBuilder<T> producerBuilder) {
 		map.from(properties::getAccessMode).to(producerBuilder::accessMode);
 	}
 
+	<T> void customizeTemplate(PulsarTemplate<T> template) {
```

Comment:
> Will bad things happen if we accidentally call customizeClientBuilder without customizeTemplate as transactions would then be partially configured?

In that case, no harm is done if you enable txn on the client but do not use them. However, If we instead called `customizeTemplate` but forgot to call `customizeClientBuilder` then the code would bark at runtime when attempting to create a transaction. 

Moving to a single customizeTransactions method would solve that but we would also have to add in the configuration for the PulsarContainerProperties which does something similar. I will see what that looks like and follow up either w/ that code change or my reasoning why not.

Reply:
Looking into this further, it gets a bit complicated trying to configure the client inline w/ the template (as well as the listener). By the time the client gets to the template (or listener) it has already been built and the builder has already been customized. Also, the client is required by all the things and as such it will always have customizeClient called.  

However, if you feel strongly about this I can spend a bit more time trying a few things.